File: jade/src/main/java/com/blockstream/jade/entities/JadeError.java
package com.blockstream.jade.entities;

public class JadeError extends RuntimeException {
    final int code;
    final Object data;

    // Local error codes
    public static final int UNSUPPORTED_FIRMWARE_VERSION = -100;
    public static final int JADE_RPC_MSG_TIMEOUT = -200;
    public static final int JADE_MSGS_OUT_OF_SYNC = -300;

    // RPC error codes
    public static final int CBOR_RPC_INVALID_REQUEST = -32600;
    public static final int CBOR_RPC_UNKNOWN_METHOD = -32601;
    public static final int CBOR_RPC_BAD_PARAMETERS = -32602;
    public static final int CBOR_RPC_INTERNAL_ERROR = -32603;

    public static final int CBOR_RPC_USER_CANCELLED = -32000;
    public static final int CBOR_RPC_PROTOCOL_ERROR = -32001;
    public static final int CBOR_RPC_HW_LOCKED = -32002;
    public static final int CBOR_RPC_NETWORK_MISMATCH = -32003;

    public JadeError(final int code, final String message, final Object data) {
        super(message);
        this.code = code;
        this.data = data;
    }

    public int getCode() {
        return code;
    }

    public Object getData() {
        return data;
    }
}

File: jade/src/commonMain/kotlin/com/blockstream/jade/data/JadeError.kt
package com.blockstream.jade.data

class JadeError constructor(val code: Int, message: String, val data: Any? = null) : Exception(message) {
    companion object {
        // Local error codes
        const val UNSUPPORTED_FIRMWARE_VERSION: Int = -100
        const val JADE_RPC_MSG_TIMEOUT: Int = -200
        const val JADE_MSGS_OUT_OF_SYNC: Int = -300

        // RPC error codes
        const val CBOR_RPC_INVALID_REQUEST: Int = -32600
        const val CBOR_RPC_UNKNOWN_METHOD: Int = -32601
        const val CBOR_RPC_BAD_PARAMETERS: Int = -32602
        const val CBOR_RPC_INTERNAL_ERROR: Int = -32603

        const val CBOR_RPC_USER_CANCELLED: Int = -32000
        const val CBOR_RPC_PROTOCOL_ERROR: Int = -32001
        const val CBOR_RPC_HW_LOCKED: Int = -32002
        const val CBOR_RPC_NETWORK_MISMATCH: Int = -32003
    }
}
